bigbluebutton/bigbluebutton
# refactor (plugins): Read plugins configs from a manifest file instead of client settings

### What does this PR do?

It changes the plugin loading flow to be from URL parameter pointing to a `manifest.json` file instead of the previous plugin settings, this will lay the grounds to create a new server-side component for plugins, which will later on be better explained (following PRs).

See demo of a sample plugin useMeeting as an example of how one can load a plugin now:

https://github.com/user-attachments/assets/daa33be5-2923-4f66-b01b-70ad2bcdf6ec

So in the video, one can see that the following plugin parameters have been given:

```
pluginsManifests=[{url:"http://172.17.0.1:4701/manifest.json", checksum="<sha256 of the file>"}]
```

Parameter `pluginsManifests` is a list of the `manifest.json` URLs and its file checksum (which can be omitted, if that's the case, then it will pass that verification - one can obtain it with the command `sha256sum manifest.json`)  and the `meta_pluginSettingsUserInformation` is a meta parameter that indicates what is going to be replaced in the `manifest.json` template strings.

```json
{
  "requiredSdkVersion": "~0.0.59",
  "name": "sampleUseMeeting",
  "javascriptEntrypointIntegrity": "aad1db35f0b1aa1072936d9977fa3cb7",
  "javascriptEntrypointUrl": "http://127.0.0.1:4701/static/SampleUseMeeting.js",
  "localesBaseUrl": "https://cdn.dominio.com/pluginabc/",
  "eventPersistence": {
      "isEnabled": true,
      "maximumPayloadSizeInBytes": 1024,
      "rateLimiting": {
          "messagesAllowedPerSecond": 10,
          "messagesAllowedPerMinute": 20
      }
  }
}
```

One can also test it with the new plugin lms-user-information, this would require the following URL parameter when testing in a dev server:

```
pluginsManifests=[{url:"http://172.17.0.1:4701/manifest.json", checksum="<sha256 of the file>"}]
meta_pluginSettingsUserInformation=https://localhost:9000/api/users
```

See the manifest of `lms-user-information` (there are some fields that will not be part of the official lms-user-information, they are there just to test if everything will be in place - that's the case for the data-channel, which is currently used as a matter of testing, as well observed in the logs):

```json
{
  "requiredSdkVersion": "~0.0.59",
  "name": "lmsUserInformation",
  "javascriptEntrypointUrl": "http://127.0.0.1:4701/static/LmsUserInformation.js",
  "localesBaseUrl": "https://cdn.dominio.com/pluginabc/",
  "eventPersistence": {
      "isEnabled": true,
      "maximumPayloadSizeInBytes": 1024,
      "rateLimiting": {
          "messagesAllowedPerSecond": 10,
          "messagesAllowedPerMinute": 20
      }
  },
  "dataChannels": [
      {
          "name": "goldStars",
          "pushPermission": ["moderator", "presenter"],
          "replaceOrDeletePermission": ["creator", "moderator"]
      }
  ],
  "remoteDataSources": [
      {
          "name": "allUsers",
          "url": "${meta_pluginSettingsUserInformation}",
          "fetchMode": "onMeetingCreate",
          "permissions": ["moderator", "viewer"]
      }
  ]
}
```

(P.S.: The error logs in the console are just happening due to the development mode of the plugin, when built, this will not happen)

### Motivation
The LMS integration.

### How to test
Just run the plugin, and send a `create` call with the URL parameters mentioned here. This will allow you to see the log of the loaded plugin (in our case, the lms-user-information) and the log of the data-channel. If you want to see the star and all the extensible areas, there's the need to run the server-side of this plugin:

In the root of the plugin, just run: 
```bash
cd server-side/
docker-compose up -d
```
It needs the docker-compose command as well as the docker itself.

### More
Closely related to PR: https://github.com/bigbluebutton/plugin-lms-user-information/pull/1
